dhall/src/Dhall/TypeCheck.hs

```diff
@@ -3140,6 +3155,47 @@ prettyTypeMessage (HandlerInputTypeMismatch expr0 expr1 expr2) =
         txt1 = insert expr1
         txt2 = insert expr2
 
+prettyTypeMessage (DependentlyTypedHandler label handlerType variable) =
+    ErrorMessages {..}
+  where
+    short = "Dependently typed handler"
+
+    long =
+        "Explanation: You can ❰merge❱ the alternatives of a union using a record with one\n\
+        \handler per alternative, like this:                                             \n\
+        \                                                                                \n\
+        \                                                                                \n\
+        \    ┌─────────────────────────────────────────────────────────────────┐         \n\
+        \    │ let union    = < Left : Natural | Right : Bool >.Left 2         │         \n\
+        \    │ let handlers = { Left = Natural/even, Right = λ(x : Bool) → x } │         \n\
+        \    │ in  merge handlers union : Bool                                 │         \n\
+        \    └─────────────────────────────────────────────────────────────────┘         \n\
+        \                                                                                \n\
+        \                                                                                \n\
+        \... but the type of the ❰merge❱ may not depend on the union value.              \n\
+        \                                                                                \n\
+        \For example, the following expression is " <> _NOT <> " valid:                  \n\
+        \                                                                                \n\
+        \                                                                                \n\
+        \    ┌────────────────────────────────────┐                                      \n\
+        \    │ let handlers = { x = None }        │                                      \n\
```

Comment:
Perhaps add an `Invalid: …` in the right-hand margin explaining why it is not valid.  For example, you can say something along the lines of: "Invalid: The output type is ❰Optional a❱, which references the input type ❰a❱"